### ISSUE TITLE
bad chunk after recovering from CrashLoopBackoff

### ISSUE THREAD

**Issue by jrockway** [pachyderm/pachyderm#6596]
I did a weird thing and got a weird result with beta3.  I am converting a workspace from beta2 to beta3 and there were some problems with my conversion -- I set WORKER_SIDECAR_IMAGE to worker instead of pachd.  This caused the pipeline to never start successfully (crashloopbackoff because k8s couldn't run /pachd).  I fixed the config, restarted, and deleted the underlying RCs.  The RCs scaled back up but failed due to Error.  This log line in `storage` seemed most relevant, but isn't the cause of the crash (the last log line was an error about not being able to start the s3 gateway, which sounds spurious to me).  I'm not actually sure what actually crashed, but I can pull logs from Stackdriver if this doesn't seem problematic but the behavior does.

```
2021-07-23T04:08:28Z ERROR pfs.API.GetFileTAR {"duration":0.026977937,"error":"bad chunk. HAVE: 30653537353163303236653534336232653861623265623036303939646161316431653564663437373738663737383766616162343563646631326665336138 WANT: 36376163353161663836383835616437343262636237336239
346363633963333533306433363165346465366631346536386433346139363766623166316338","request":{"file":{"commit":{"branch":{"repo":{"name":"edges","type":"spec"},"name":"master"},"id":"3f72a6752ea746b99f43ddf6ddfefcd6"},"path":"spec"}},"stack":["github.com/pachyderm/pachyderm/v2/src/in
ternal/storage/chunk.verifyData\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/chunk/transform.go:193","github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk.Get.func1\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage
/chunk/transform.go:58","github.com/pachyderm/pachyderm/v2/src/internal/storage/kv.(*objectAdapter).Get.func1\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/kv/obj_adapter.go:42","github.com/pachyderm/pachyderm/v2/src/internal/storage/kv.(*objectAdapt
er).withBuffer\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/kv/obj_adapter.go:68","github.com/pachyderm/pachyderm/v2/src/internal/storage/kv.(*objectAdapter).Get\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/kv/obj_ada
pter.go:35","github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk.(*trackedClient).Get\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/chunk/client.go:129","github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk.Get\n\t/Users/avigil/Pachy
derm/Pachyderm_releases/pachyderm/src/internal/storage/chunk/transform.go:57","github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk.(*DataReader).Get\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/chunk/reader.go:74","github.com/pachyderm/pach
yderm/v2/src/internal/storage/chunk.(*Reader).Get.func1\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/chunk/reader.go:46","github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk.(*Reader).Iterate\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pa
chyderm/src/internal/storage/chunk/reader.go:33","github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk.(*Reader).Get\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/chunk/reader.go:45","github.com/pachyderm/pachyderm/v2/src/internal/storage/fil
eset/index.(*levelReader).setup\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/fileset/index/reader.go:164","github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index.(*levelReader).Read\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyde
rm/src/internal/storage/fileset/index/reader.go:143","io.ReadAtLeast\n\t/usr/local/go/src/io/io.go:328","io.ReadFull\n\t/usr/local/go/src/io/io.go:347","encoding/binary.Read\n\t/usr/local/go/src/encoding/binary/binary.go:166","github.com/pachyderm/pachyderm/v2/src/internal/pbutil.
(*readWriter).ReadBytes\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/pbutil/pbutil.go:39","github.com/pachyderm/pachyderm/v2/src/internal/pbutil.(*readWriter).Read\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/pbutil/pbutil.go:57","gi
thub.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index.(*Reader).Iterate\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/fileset/index/reader.go:51","github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset.(*Reader).Iterate\n\t/Users
/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/fileset/reader.go:48","github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset.(*Storage).Size\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/fileset/storage.go:273","github.co
m/pachyderm/pachyderm/v2/src/server/pfs/server.(*driver).commitSize\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/server/pfs/server/driver_file.go:517","github.com/pachyderm/pachyderm/v2/src/server/pfs/server.(*driver).inspectCommit\n\t/Users/avigil/Pachyderm/Pachyde
rm_releases/pachyderm/src/server/pfs/server/driver.go:980","github.com/pachyderm/pachyderm/v2/src/server/pfs/server.(*driver).openCommit\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/server/pfs/server/driver_file.go:128","github.com/pachyderm/pachyderm/v2/src/server/
pfs/server.(*driver).getFile\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/server/pfs/server/driver_file.go:176","github.com/pachyderm/pachyderm/v2/src/server/pfs/server.(*apiServer).GetFileTAR.func3\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/server/
pfs/server/api_server.go:514","github.com/pachyderm/pachyderm/v2/src/internal/storage/metrics.ReportRequestWithThroughput.func1\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/metrics/metrics.go:68","github.com/pachyderm/pachyderm/v2/src/internal/stora
ge/metrics.ReportRequest\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/internal/storage/metrics/metrics.go:54","github.com/pachyderm/pachyderm/v2/src/internal/storage/metrics.ReportRequestWithThroughput\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/inte
rnal/storage/metrics/metrics.go:67","github.com/pachyderm/pachyderm/v2/src/server/pfs/server.(*apiServer).GetFileTAR\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/server/pfs/server/api_server.go:512","github.com/pachyderm/pachyderm/v2/src/server/pfs/server.(*validate
dAPIServer).GetFileTAR\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/server/pfs/server/val_server.go:154","github.com/pachyderm/pachyderm/v2/src/pfs._API_GetFileTAR_Handler\n\t/Users/avigil/Pachyderm/Pachyderm_releases/pachyderm/src/pfs/pfs.pb.go:4951"]}
```

**Comment by jdoliner** [comment]
@brycemcanally would you mind taking a look at this? The things @jrockway did aren't likely to be done by a user but the fact that it could cause a bad chunk is a little concerning.

**Comment by brycemcanally** [comment]
The bad chunk error is just a sanity check in the storage layer to ensure that data isn't changing underneath us. The `30653537353163303236653534336232653861623265623036303939646161316431653564663437373738663737383766616162343563646631326665336138` hash is for empty bytes, which likely means that object storage got out of sync with what we track in Postgres. I don't think we are propagating gets that fail due to non-existence correctly, which would explain why this is the error you see. I saw the same error when we weren't mounting in the correct volume for local deploys. @jrockway did you do anything that might have affected either the object storage client setup or object storage state?

**Comment by jrockway** [comment]
I don't think I did anything that would affect the object storage setup.  The workspace did start out as a beta2 workspace, but no data was added while beta2 was running (but, schema migrations etc. happened on beta2).  Data was probably added while the worker image tag was wrong, though.

pachd could have died at an inopportune time.  Hopefully that can't result in missing data.

I wonder if we can make a more isolated test case for this -- set the worker images to something that will crash upon startup, add data, restart pachd with valid images, see if this bug comes back.
